tun/tun_illumos.go

```diff
@@ -0,0 +1,381 @@
+		events:      make(chan Event, 10),
+		errors:      make(chan error, 1),
+	}
+
+	err = tun.configureSTREAMS()
+	if err != nil {
+		tun.Close()
+		return nil, err
+	}
+
+	go tun.routineRouteListener()
+
+	err = tun.setMTU(mtu)
+	if err != nil {
+		tun.Close()
+		return nil, err
+	}
+
+	// The interface up message was emitted before the listener was started
+	// when the tun device was created. Emit the event here to compensate.
+	tun.events <- EventUp
+
+	// Always send an MTUUpdate event to ensure correct detection at startup
+	tun.events <- EventMTUUpdate
+
+	return tun, nil
+}
+
+func (tun *NativeTun) configureSTREAMS() error {
+	// get back the ppa number from the name
+	var ppa int
+	_, err := fmt.Sscanf(tun.name, "tun%d", &ppa)
+	if err != nil {
+		return fmt.Errorf("received a bogus tun name: %s", tun.name)
+	}
+
+	// Unclear why, but we need to open a second handle on the tun device for
+	// the ioctls that follow. Reusing the existing one does not work.
+	tunFD, err := unix.Open(tunNode, unix.O_RDWR, 0)
+	if err != nil {
+		return fmt.Errorf("Could not open TUN (%s)", tunNode)
+	}
+	defer unix.Close(tunFD)
+
+	// Push the IP module onto the TUN device.
+	if err = unix.IoctlSetString(tunFD, unix.I_PUSH, "ip"); err != nil {
+		return fmt.Errorf("unable to push IP module: %v", err)
+	}
+
+	// select which ppa we're using
+	req := int(unix.IF_UNITSEL)
+	err = unix.IoctlSetPointerInt(tunFD, uint(req), ppa)
+	if err != nil {
+		return fmt.Errorf("unable to IF_UNITSEL: %v", err)
+	}
+
+	// link the tun stream to the IP multiplexor
+	// If we used I_PLINK the connection would persist after we close the ipfd
+	// which causes cleanup issues. By using I_LINK and holding ipfd open,
+	// everything gets cleaned up nicely when we close the file descriptors
+	// or when the process terminates.
+	muxid, err := unix.IoctlSetIntRetInt(tun.ipfd, unix.I_LINK, tunFD)
+	if err != nil {
+		return fmt.Errorf("unable to I_LINK: %v", err)
+	}
+
+	var l unix.Lifreq
+	if err := l.SetName(tun.name); err != nil {
+		return fmt.Errorf("failed to set name on Lifreq :%v", err)
+	}
+
+	// set the IP muxid
+	reqnum := int(unix.SIOCSLIFMUXID)
+	l.SetLifruInt(muxid)
+	if err = unix.IoctlLifreq(tun.ipfd, uint(reqnum), &l); err != nil {
+		return fmt.Errorf("unable to SIOCSLIFMUXID: %v", err)
+	}
+
+	return nil
+}
+
+func (tun *NativeTun) Name() (string, error) {
+	return tun.name, nil
+}
+
+func (tun *NativeTun) File() *os.File {
+	return tun.tunFile
+}
+
+func (tun *NativeTun) Events() chan Event {
+	return tun.events
+}
+
+func (tun *NativeTun) Read(buf []byte, offset int) (int, error) {
+	select {
+	case err := <-tun.errors:
+		return 0, err
+	default:
+		_, read, _, err := unix.Getmsg(tun.fd, nil, buf[offset:])
+		if err != nil {
+			return 0, err
+		}
+		return len(read), nil
+	}
+}
+
+func (tun *NativeTun) Write(buf []byte, offset int) (int, error) {
+	if err := unix.Putmsg(tun.fd, nil, buf[offset:], 0); err != nil {
+		return 0, err
+	}
+	return len(buf), nil
```

Comment:
Should this be `len(buf[offset:])`?
(If `offset` is always `0`, it is fine, but it could be otherwise, right?)

Reply:
Nice catch. It should indeed. 

You can make the first line of the function `buf = buf[offset:]`, and then just deal with buf itself the rest of the way through.

Reply:
Fixed.